src/java.base/share/classes/java/math/BigInteger.java

```diff
@@ -3991,12 +3993,9 @@ final int compareMagnitude(BigInteger val) {
             return -1;
         if (len1 > len2)
             return 1;
-        for (int i = 0; i < len1; i++) {
-            int a = m1[i];
-            int b = m2[i];
-            if (a != b)
-                return ((a & LONG_MASK) < (b & LONG_MASK)) ? -1 : 1;
-        }
+        int i = ArraysSupport.mismatch(m1, m2, len1);
+        if (i != -1)
+            return ((m1[i] & LONG_MASK) < (m2[i] & LONG_MASK)) ? -1 : 1;
```

Comment:
As suggested earlier, you could make us of `Integer.compareUnsigned`, which is an intrinsic not involving `long`s.